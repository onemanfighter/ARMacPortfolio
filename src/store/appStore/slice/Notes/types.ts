export type NotesState = {};

export interface NotesAppAction {}

export type NotesStateSlice = NotesState & NotesAppAction;
