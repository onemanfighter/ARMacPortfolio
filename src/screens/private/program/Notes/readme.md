# Notes

The notes screen is a simple text editor that allows the user to write notes about the program. The notes are saved to the database and can be viewed and edited at any time.

## Components

Used code editor: https://draftjs.org/
