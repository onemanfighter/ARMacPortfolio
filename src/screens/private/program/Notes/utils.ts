const getFirstLineFromString = (str: string) => {
  return str.split('\n')[0];
};

const getAllStringExceptFirstLine = (str: string) => {
  return str.split('\n').slice(1).join('\n');
};

export { getFirstLineFromString, getAllStringExceptFirstLine };
