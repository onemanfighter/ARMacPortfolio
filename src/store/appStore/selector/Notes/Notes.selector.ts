import { AppStoreState } from '../../appStore';

const notesSelector = (state: AppStoreState) => ({
  notes: state.Notes.notes,
  selectedNote: (id: string) => state.Notes.notes[id],
  addNote: state.Notes.addNote,
  deleteNote: state.Notes.deleteNote,
  editNote: state.Notes.editNote,
});

export { notesSelector };
