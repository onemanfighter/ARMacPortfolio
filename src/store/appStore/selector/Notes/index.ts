export { notesSelector } from './Notes.selector';
