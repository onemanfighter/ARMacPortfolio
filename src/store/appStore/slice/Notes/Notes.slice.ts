import { AppStoreSlice } from '../../appStore';
import { NotesState, NotesStateSlice } from './types';

const defaultNotesState: NotesState = {
  notes: {},
};

const createNotesSlice: AppStoreSlice<NotesStateSlice> = (set) => ({
  ...defaultNotesState,
  addNote: (note) =>
    set((state) => {
      state.Notes.notes[note.id] = note;
    }),
  deleteNote: (id) =>
    set((state) => {
      delete state.Notes.notes[id];
    }),
  editNote: (newNote) =>
    set((state) => {
      state.Notes.notes[newNote.id] = newNote;
    }),
});

export default createNotesSlice;
