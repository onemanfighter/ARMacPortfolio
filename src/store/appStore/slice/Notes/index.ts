export { default as createNotesSlice } from './Notes.slice';
export type { NotesStateSlice, NotesState, Note } from './types';
