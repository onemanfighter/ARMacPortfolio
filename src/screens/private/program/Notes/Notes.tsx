import { ContentState, Editor, EditorState } from 'draft-js';
// import 'draft-js/dist/Draft.css';
import { Box, Card, IconButton, Text } from '@chakra-ui/react';

import { NotesProps } from './type';
import { appStore, notesSelector, useShallow } from '@appStore';
import { useEffect, useState } from 'react';
import { getAllStringExceptFirstLine, getFirstLineFromString } from './utils';

const color = '#cc9900';
const Notes = (props: NotesProps) => {
  const [editorState, setEditorState] = useState(() =>
    EditorState.createEmpty(),
  );
  const [selectedNoteId, setSelectedNoteId] = useState<string>('0');

  const { notes, addNote, selectedNote, getCurrentId, deleteNote, editNote } =
    appStore(useShallow(notesSelector));
  const currentNote = selectedNote(selectedNoteId);

  useEffect(() => {
    if (selectedNoteId !== '0') {
      setEditorState(
        EditorState.createWithContent(
          ContentState.createFromText(
            selectedNoteId
              ? currentNote.title + '\n' + currentNote.description
              : '',
          ),
        ),
      );
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedNoteId]);

  return (
    <Box
      width={'100%'}
      height={'100%'}
      display={'flex'}
      flexDir={'column'}
      aria-label="notes-app"
    >
      <Box
        width={'100%'}
        height={'6%'}
        bg={'#332600'}
        borderBottom={'1px solid #000000'}
        display={'flex'}
        flexDir={'row'}
        gap={2}
      >
        <IconButton
          aria-label="Add Note"
          size={'sm'}
          icon={<Text fontSize={20}>+</Text>}
          onClick={() =>
            addNote({
              id: (getCurrentId() + 1) as unknown as string,
              title: 'Title',
              description: 'Description',
              date: new Date().toLocaleDateString(),
            })
          }
        />
        <IconButton
          aria-label="Delete Note"
          size={'sm'}
          icon={<Text fontSize={20}>-</Text>}
          onClick={() => {
            if (selectedNoteId !== '0') {
              deleteNote(selectedNoteId);
              setSelectedNoteId('0');
            }
          }}
        />
      </Box>
      <Box
        width={'100%'}
        height={'94%'}
        bg={'#332600'}
        borderBottom={'1px solid #000000'}
        display={'flex'}
        flexDir={'row'}
      >
        <Box
          width={'20%'}
          height={'100%'}
          bg={'#1a13009f'}
          borderRight={'1px solid #000000'}
          gap={0}
        >
          {Object.values(notes).map((note) => (
            <>
              <Card
                key={note.id}
                bg={selectedNoteId === note.id ? color : '#1a13009f'}
                margin={2}
                padding={1}
                px={3}
                onClick={() => {
                  if (selectedNoteId !== '0') {
                    editNote({
                      id: selectedNoteId,
                      title: getFirstLineFromString(
                        editorState.getCurrentContent().getPlainText(),
                      ),
                      description: getAllStringExceptFirstLine(
                        editorState.getCurrentContent().getPlainText(),
                      ),
                      date: selectedNoteId ? currentNote?.date : '',
                    });
                  }
                  setSelectedNoteId(note.id);
                }}
                _hover={{
                  cursor: 'pointer',
                }}
                color={'#ffffff'}
                borderBottom={'1px solid #000000'}
              >
                <Text variant="h5" noOfLines={1}>
                  {note.title}
                </Text>
                <Text fontSize={14}>
                  <Text>{note.date}</Text>
                  {note.description}
                </Text>
              </Card>
            </>
          ))}
        </Box>
        <Box
          width={'80%'}
          height={'100%'}
          padding={3}
          bg={'#1a1a1a'}
          overflowY={'auto'}
          color={'white'}
        >
          <Editor editorState={editorState} onChange={setEditorState} />
        </Box>
      </Box>
    </Box>
  );
};

export default Notes;
