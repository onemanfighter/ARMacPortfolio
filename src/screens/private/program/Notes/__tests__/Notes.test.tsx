import { fireEvent, render, screen } from '@testing-library/react';
import Notes from '../Notes';
import { renderHook } from '@testing-library/react-hooks';
import { appStore } from '@appStore';

describe('Notes', () => {
  it('should render correctly to match snapshot', () => {
    const { container } = render(<Notes />);

    expect(container).toMatchSnapshot();
  });

  it('should invoke add note on clicking add note button', () => {
    const { result } = renderHook(() => appStore());
    const { container } = render(<Notes />);

    fireEvent.click(screen.getByLabelText('add-note'));

    expect(container).toMatchSnapshot();
    expect(result.current.Notes.notes['1']).toEqual({
      id: '1',
      title: 'Title',
      description: 'Description',
      date: new Date().toLocaleDateString(),
    });
  });

  it('should invoke delete note on clicking delete note button', () => {
    const { result } = renderHook(() => appStore());
    const { container } = render(<Notes />);

    fireEvent.click(screen.getByLabelText('add-note'));
    fireEvent.click(screen.getByLabelText('note-card-1'));
    fireEvent.click(screen.getByLabelText('delete-note'));

    expect(container).toMatchSnapshot();
    expect(result.current.Notes.notes['1']).toBeUndefined();
  });

  it('should select and deselect on clicking on card', () => {
    const { result } = renderHook(() => appStore());
    const { container } = render(<Notes />);

    fireEvent.click(screen.getByLabelText('add-note'));
    fireEvent.click(screen.getByLabelText('add-note'));

    fireEvent.click(screen.getByLabelText('note-card-2'));
    fireEvent.click(screen.getByLabelText('note-card-1'));

    // Card 1 is selected
    expect(container).toMatchSnapshot();
  });
});
