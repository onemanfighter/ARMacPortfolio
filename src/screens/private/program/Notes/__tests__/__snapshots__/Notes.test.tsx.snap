// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes should invoke add note on clicking add note button 1`] = `
<div>
  <div
    aria-label="notes-app"
    class="css-1v8gcgw"
  >
    <div
      class="css-cum3m2"
    >
      <button
        aria-label="add-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          +
        </p>
      </button>
      <button
        aria-label="delete-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          -
        </p>
      </button>
    </div>
    <div
      class="css-1fjg3y4"
    >
      <div
        class="css-1culoml"
      >
        <div
          aria-label="note-card-1"
          class="chakra-card css-6capty"
        >
          <p
            class="chakra-text css-1i664ua"
          >
            Title
          </p>
          <p
            class="chakra-text css-d050u0"
          >
            5/14/2020
            <p
              class="chakra-text css-116pnhr"
            >
               -&gt; 
              Description
            </p>
          </p>
        </div>
      </div>
      <div
        class="css-1xb7lh4"
      >
        <div>
          Editor
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notes should invoke delete note on clicking delete note button 1`] = `
<div>
  <div
    aria-label="notes-app"
    class="css-1v8gcgw"
  >
    <div
      class="css-cum3m2"
    >
      <button
        aria-label="add-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          +
        </p>
      </button>
      <button
        aria-label="delete-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          -
        </p>
      </button>
    </div>
    <div
      class="css-1fjg3y4"
    >
      <div
        class="css-1culoml"
      />
      <div
        class="css-1xb7lh4"
      >
        <div>
          Editor
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notes should render correctly to match snapshot 1`] = `
<div>
  <div
    aria-label="notes-app"
    class="css-1v8gcgw"
  >
    <div
      class="css-cum3m2"
    >
      <button
        aria-label="add-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          +
        </p>
      </button>
      <button
        aria-label="delete-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          -
        </p>
      </button>
    </div>
    <div
      class="css-1fjg3y4"
    >
      <div
        class="css-1culoml"
      />
      <div
        class="css-1xb7lh4"
      >
        <div>
          Editor
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notes should select and deselect on clicking on card 1`] = `
<div>
  <div
    aria-label="notes-app"
    class="css-1v8gcgw"
  >
    <div
      class="css-cum3m2"
    >
      <button
        aria-label="add-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          +
        </p>
      </button>
      <button
        aria-label="delete-note"
        class="chakra-button css-p7mkev"
        type="button"
      >
        <p
          aria-hidden="true"
          class="chakra-text css-bk9fzy"
          focusable="false"
        >
          -
        </p>
      </button>
    </div>
    <div
      class="css-1fjg3y4"
    >
      <div
        class="css-1culoml"
      >
        <div
          aria-label="note-card-1"
          class="chakra-card css-mg41zh"
        >
          <p
            class="chakra-text css-1i664ua"
          >
            Title
          </p>
          <p
            class="chakra-text css-d050u0"
          >
            5/14/2020
            <p
              class="chakra-text css-116pnhr"
            >
               -&gt; 
              Description
            </p>
          </p>
        </div>
        <div
          aria-label="note-card-2"
          class="chakra-card css-6capty"
        >
          <p
            class="chakra-text css-1i664ua"
          >
            Title
          </p>
          <p
            class="chakra-text css-d050u0"
          >
            5/14/2020
            <p
              class="chakra-text css-116pnhr"
            >
               -&gt; 
              Description
            </p>
          </p>
        </div>
      </div>
      <div
        class="css-1xb7lh4"
      >
        <div>
          Editor
        </div>
      </div>
    </div>
  </div>
</div>
`;
